### ISSUE TITLE
terminal weirdness - uparrow at bottom line - doesn't redraw command.

### ISSUE THREAD

**Issue by Dunbaratu** [KSP-KOS/KOS#1289]
To reproduce:

- Run some commands interactively until the cursor scrolls far enough down that it has caused the terminal to scroll some, and the cursor is on the bottom row.

- Use up-arrow to retrieve the previous command.

The command is invisible but it is in the buffer.  If you hit enter it will run again.  You can edit it, but you have to edit it blindly - i.e. backspace some and then retype stuff.

The moment you make an edit that will insert a new character into the line, THEN it redraws the line.  Before that it does not.  Deleting a char doesn't redraw it - only adding a new nonblank character does.

**Comment by Dunbaratu** [comment]
I am suspecting the problem is in the "dirty check" that's used to decide whether or not there's a need to draw those characters.  Note the problem happens the same on the in-game window and in telnet.

**Comment by Dunbaratu** [comment]
Interesting - it's dumping an exception silently to the output log when this happens.  The exception isn't causing it to barf - it's just silently continuing on.
```
ArgumentOutOfRangeException: Argument is out of range.
Parameter name: index
  at System.Collections.Generic.List`1[kOS.Safe.Screen.IScreenBufferLine].get_Item (Int32 index) [0x00000] in <filename unknown>:0 
  at kOS.Safe.Screen.ScreenBuffer.MarkRowsDirty (Int32 startRow, Int32 numRows) [0x00000] in <filename unknown>:0 
  at kOS.Screen.Interpreter.ShowCommandHistoryEntry (Int32 deltaIndex) [0x00000] in <filename unknown>:0 
  at kOS.Screen.Interpreter.SpecialKey (Char key) [0x00000] in <filename unknown>:0 
  at kOS.AddOns.RemoteTech.RemoteTechInterpreter.SpecialKey (Char key) [0x00000] in <filename unknown>:0 
  at kOS.Screen.TermWindow.SpecialKey (Char key) [0x00000] in <filename unknown>:0 
  at kOS.Screen.TermWindow.ProcessOneInputChar (Char ch, kOS.UserIO.TelnetSingletonServer whichTelnet) [0x00000] in <filename unknown>:0 
  at kOS.Screen.TermWindow.ProcessKeyEvents () [0x00000] in <filename unknown>:0 
  at kOS.Screen.TermWindow.OnGUI () [0x00000] in <filename unknown>:0 
 ```

**Comment by hvacengi** [comment]
I seem to remember a similar error in a recent issue (again related to the line buffer for the terminal) but I can't seem to find it.  I want to say that you @Dunbaratu fixed it, maybe you recall?

**Comment by Dunbaratu** [comment]
I think there was an earlier problem with it throwing an exception in some edge cases of the screen redraw, but it was fixed - this must be being caused by a different set of circumstances.

**Comment by Dunbaratu** [comment]
Note to self to try when I come back to this:

I think the exception is caused by the fact that in Interpreter.cs, in the following lines:
```
line 112]                    MarkRowsDirty(LineSubBuffer.PositionRow, LineSubBuffer.RowCount);
line 113]                    LineSubBuffer.Wipe();
line 114]                    UpdateLineSubBuffer();
```
The `MarkRowsDirty` call blows past the last line of the buffer (giving the out of bounds error) because the subbuffer starts on the last row and extends past the end of the screen buffer.  The call to `UpdateLineSubBuffer` on line 114 normally would cause the extra line to be appended to the buffer (eventually, when it finally gets to the call to `KeepCursorInBounds` nested inside it), but that doesn't occur until after it attempted `MarkRowsDirty`.

The problem is I have to solve this chicken-and-egg situation by exploding some of the parts of these out.  At the moment, `MarkRowsDirty` must be called prior to `UpdateLineSubBuffer` and `UpdateLineSubBuffer` must be called prior to `MarkRowsDIrty`.

**Comment by Dunbaratu** [comment]
Closed #730 from this too.
